### PR TITLE
Allow bypassing temporary table during insert for jdbc-based connectors

### PR DESCRIPTION
**Pull request by wendigo**

### CODE REVIEW

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/BaseJdbcClient.java [hunk 1]
```diff
@@ -683,6 +696,11 @@ protected void renameTable(ConnectorSession session, String catalogName, String
     @Override
     public void finishInsertTable(ConnectorSession session, JdbcOutputTableHandle handle)
     {
+        if (isNonTransactionalInsert(session)) {
+            checkState(handle.getTemporaryTableName().equals(handle.getTableName()), "Could not finish non-transactional insert when temporary table was used");
```
Comment (losipiuk) [inline plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/BaseJdbcClient.java:700]: `Unexpected use of temporary table when non transactional inserts are enabled`?

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcMetadataSessionProperties.java [hunk 2]
```diff
@@ -38,6 +38,7 @@
     public static final String TOPN_PUSHDOWN_ENABLED = "topn_pushdown_enabled";
     public static final String DOMAIN_COMPACTION_THRESHOLD = "domain_compaction_threshold";
     public static final String INSERT_BATCH_SIZE = "insert_batch_size";
+    public static final String NON_TRANSACTIONAL_INSERT = "non_transactional_insert";
```
Comment (losipiuk) [inline plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcMetadataSessionProperties.java:41]: Can we make property/config name more self descriptive.
Maybe negate it and name `use_temporary_table_on_insert`?

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/BaseJdbcClient.java [hunk 3]
```diff
@@ -59,13 +59,15 @@
 import java.util.function.Function;
 
 import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
```
Comment (losipiuk) [inline plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/BaseJdbcClient.java:62]: Put some motivation/descritpion in commit message.

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcMetadataSessionProperties.java (hunk 2 above)
Comment (hashhar) [reply plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcMetadataSessionProperties.java:41]: This is a unsafe feature so something which conveys that seems useful to me.

File: plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/BaseJdbcConnectorTest.java [hunk 4]
```diff
@@ -1262,6 +1261,28 @@ public void testDeleteWithVarcharPredicate()
...
 
+    @Test
+    public void testInsertWithoutTemporaryTable()
+    {
+        if (!hasBehavior(SUPPORTS_CREATE_TABLE)) {
+            throw new SkipException("CREATE TABLE is required for testing non-transactional write support");
+        }
+        Session session = Session.builder(getSession())
+                .setCatalogSessionProperty(getSession().getCatalog().orElseThrow(), "non_transactional_insert", "true")
+                .setCatalogSessionProperty(getSession().getCatalog().orElseThrow(), "insert_batch_size", "16")
```
Comment (hashhar) [inline plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/BaseJdbcConnectorTest.java:1272]: Does this matter? IIUC `beginInsertTable` always uses direct insert now if session property is set.

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcMetadataSessionProperties.java (hunk 2 above)
Comment (ssheikin) [reply plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcMetadataSessionProperties.java]: Please rollback the change to `NON_TRANSACTIONAL_INSERT`

File: plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/BaseJdbcConnectorTest.java [hunk 5]
```diff
@@ -1262,6 +1261,27 @@ public void testDeleteWithVarcharPredicate()
...
+            throw new SkipException("CREATE TABLE is required for testing non-transactional write support");
+        }
+        Session session = Session.builder(getSession())
+                .setCatalogSessionProperty(getSession().getCatalog().orElseThrow(), "use_temporary_table_on_insert", "false")
+                .build();
+
+        int numberOfRows = 1000;
+        try (TestTable table = new TestTable(
+                getQueryRunner()::execute,
+                "test_insert_batch_size",
```
Comment (ssheikin) [inline plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/BaseJdbcConnectorTest.java:1277]: rename

File: plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/BaseJdbcConnectorTest.java [hunk 6]
```diff
@@ -1262,6 +1261,27 @@ public void testDeleteWithVarcharPredicate()
...
+    public void testInsertWithoutTemporaryTable()
+    {
+        if (!hasBehavior(SUPPORTS_CREATE_TABLE)) {
+            throw new SkipException("CREATE TABLE is required for testing non-transactional write support");
+        }
+        Session session = Session.builder(getSession())
+                .setCatalogSessionProperty(getSession().getCatalog().orElseThrow(), "use_temporary_table_on_insert", "false")
+                .build();
+
+        int numberOfRows = 1000;
```
Comment (ssheikin) [inline plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/BaseJdbcConnectorTest.java:1274]: why 1000?

Comment (wendigo) [reply plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/BaseJdbcConnectorTest.java]: It's irrelevant but I've raised it to 5000 to it's larger than the default batch size insert.

**Comment by wendigo** [comment]
@hashhar @ssheikin PTAL

**Review by hashhar — APPROVED** [review]

**Review by ssheikin — APPROVED** [review]

**Comment by wendigo** [comment]
CI hit: https://github.com/trinodb/trino/issues/8477

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcMetadataConfig.java [hunk 7]
```diff
@@ -45,6 +45,9 @@
 
     private int insertBatchSize = 1000;
 
+    // Do not create temporary table during insert which ensures that all the rows will be inserted atomically.
```
Comment (hashhar) [inline plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcMetadataConfig.java:48]: Misleading comment - enabling this makes the process non-atomic.

```suggestion
    // Do not create temporary table during insert. This means that the write operation can fail and leave the table in an inconsistent state.
```

Maybe copy this to the `ConfigDescription` too?

Comment (ssheikin) [reply plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcMetadataConfig.java:48]: I had the same feeling.

Comment (wendigo) [reply plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcMetadataConfig.java]: Updated.

**Comment by wendigo** [comment]
@hashhar can we merge it?

**Review by hashhar — APPROVED** [review]
LGTM

**Comment by wendigo** [comment]
Failure is again: #8477

**Comment by hashhar** [comment]
Let's make sure when/if this gets documented there are big bold warnings about the possible ways this can fail.

**Pull request merged**